rfd/0191-workload-id-config-ux.md

```diff
@@ -0,0 +1,868 @@
+// authoirzation rules. is a resource that represents the configuration of a trust
+// domain federation.
+message WorkloadIdentity {
+  // The kind of resource represented.
+  string kind = 1;
+  // Differentiates variations of the same kind. All resources should
+  // contain one, even if it is never populated.
+  string sub_kind = 2;
+  // The version of the resource being represented.
+  string version = 3;
+  // Common metadata that all resources share.
+  teleport.header.v1.Metadata metadata = 4;
+  // The configured properties of the WorkloadIdentity
+  WorkloadIdentitySpec spec = 5;
+}
+
+// WorkloadIdentityRules holds the allow and deny authorization rules for the
+// WorkloadIdentitySpec.
+//
+// Deny rules take precedence over allow rules.
+message WorkloadIdentityRules {
+  // Allow is a list of rules, each containing a set of attribute matchers.
+  // For a rule to match, all matchers within the rule must match.
+  // If rules are specified, then at least one rule must match for issuance to
+  // be permitted.
+  // If no rules are specified, issuance is permitted.
+  repeated map<string, string> allow = 1;
+  // Deny is a set of rules, each containing a set of attribute matchers.
+  // For a rule to match, all matchers within the rule must match.
+  // If rules are specified, then all rules must not match for issuance to be
+  // permitted.
+  // If no rules are specified, issuance is permitted.
+  repeated map<string, string> deny = 2;
+}
+
+// WorkloadIdentitySPIFFE holds configuration for the issuance of
+// SPIFFE-compatible workload identity credentials when this WorkloadIdentity
+// is used.
+message WorkloadIdentitySPIFFE {
+  // Id is the path of the SPIFFE ID that will be issued in workload identity
+  // credentials when this WorkloadIdentity is used. It can be templated using
+  // attributes.
+  //
+  // Examples:
+  // - `/no/templating/used` -> `spiffe://example.teleport.sh/no/templating/used`
+  // - `/gitlab/{{ join.gitlab.project_path }}` -> `spiffe://example.teleport.sh/gitlab/org/project`
+  string id = 1; 
+}
+
+// WorkloadIdentitySpec holds the configuration element of the WorkloadIdentity
+// resource.
+message WorkloadIdentitySpec {
+  // Rules holds the authorization rules that must pass for this
+  // WorkloadIdentity to be used. See [WorkloadIdentityRules] for further
+  // documentation.
+  WorkloadIdentityRules rules = 1;
+  // SPIFFE holds configuration for the structure of SPIFFE-compatible 
+  // workload identity credentials for this WorkloadIdentity. See 
+  // [WorkloadIdentitySPIFFE] for further documentation.
+  WorkloadIdentitySPIFFE spiffe = 2;
+}
+```
+
+As per RFD 153, CRUD RPCs will be included for the WorkloadIdentity resource.
+The ability to execute these RPCs will be governed by the standard verbs with a 
+noun of `workload_identity`.
+
+The proto specification of the RPCs is omitted for conciseness.
+
+### Propagating Join Attributes into X509 Certificates
```

Comment:
I'm slightly doubtful that storing the attribute set as a `map<string, string>` may be problematic, even if it makes rule evaluation/templating simpler. I can see a few problems:

- There's attributes which will be non-string at source, that we will need to cast to a string. This may not be a huge problem but might make handling booleans more difficult, especially if we ever add predicate statement support.
- Storing a big map of repetitive keys is not fantastic in terms of efficiency, and we'll want to encode at least part of the attribute set into X509 certs.
- It'll be more difficult to detect "invalid" configurations since there won't be a defined schema.

One thought in my mind is to just handle/store attribute sets as a properly typed and nested protobuf message, but I have a few other concerns with this:

- It'll be much more challenging to write a rule/template engine that handles navigating/reflecting through the protobuf than it will be to write a rule/template engine that handles a simple map<string, string>.
  - This could be ameliorated by transporting the attributes in a proto, and having a method which then returns the flattened json of the proto. 
- It introduces some versioning compat concerns. If a new attribute is added to workload attestation in `tbot`, then the Auth Service must also be updated to a version that knows this attribute.

Reply:
As I think more on this, I do feel like having some kind of protobuf declaration of this will be useful. Even if it's not leveraged within the rules/template engine, it does provide an empirical "source of truth" for documentation purposes.

Reply:
I would also expect dots in variable names (as in a key `"join.gitlab.ref_type"`) to interfere with our existing templating/predicate engine. They tend to expect to traverse some sort of actual nested object structure.

(Map syntax would probably work, but is ugly, e.g. `attrs["join.gitlab.ref_type"]`)

Reply:
Yup - ran into this in my spike earlier. I think my mind is pretty set on using some kind of properly nested structure now.